### PR TITLE
✨ Added App Insights telemetry for calls as defined in ticket 743

### PR DESCRIPTION
**Pull request by chris-faulkner-mojdt**
Followed the court case service example (minus request based properties for user / client).

### CODE REVIEW

**Review by sarahyjja — APPROVED** [review]
It's hard for me to really verifying as I don't know Java, but it's looks like a lot of addition more than modifications, so it's looks good to me ^^

File: src/main/java/uk/gov/justice/probation/courtcasematcher/application/TelemetryConfig.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        @Override
+        public boolean matches(@NonNull ConditionContext context, @NonNull AnnotatedTypeMetadata metadata) {
+            return StringUtils.isEmpty(context.getEnvironment().getProperty("application.insights.ikey"));
+        }
+    }
+
+    @Bean
+    @Profile("!test")
+    @RequestScope
+    public Map<String, String> requestProperties(){
```
Comment (jhevans) [inline src/main/java/uk/gov/justice/probation/courtcasematcher/application/TelemetryConfig.java:40]: Don't think the matcher will need this bean as it doesn't have a REST API as such so we can probably remove it. This was only used for getting the user details in the court-case-service and being RequestScoped wouldn't be available at runtime in any event.

File: src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/MessageProcessorTest.java [hunk 2]
```diff
@@ -132,6 +121,13 @@ void givenTwoExistingAndOneNew_whenValidMessageReceivedWithMultipleSessions_Then
         verify(eventBus, timeout(1000)).post(argThat(CourtCaseUpdateEventMatcher.builder().caseNo("1600032953").build()));
         verify(eventBus, timeout(1000)).post(argThat(CourtCaseUpdateEventMatcher.builder().caseNo("1600032979").build()));
         verify(eventBus, timeout(1000)).post(argThat(CourtCaseMatchEventMatcher.builder().caseNo("1600032952").build()));
+        verify(telemetryService).trackCourtListEvent(argThat(getInfoMatcher(LocalDate.of(2020, Month.FEBRUARY, 20))));
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/MessageProcessorTest.java:124]: `argThat(matchesInfoWith(LocalDate....)` might read better here

File: src/test/java/uk/gov/justice/probation/courtcasematcher/service/TelemetryServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            .build();
+
+        telemetryService.trackOffenderMatchEvent(COURT_CODE, CASE_NO, response);
+
+        verify(telemetryClient).trackEvent(eq("PiCOffenderExactMatch"), propertiesCaptor.capture(), eq(Collections.emptyMap()));
+
+        Map<String, String> properties = propertiesCaptor.getValue();
+        assertThat(properties).hasSize(6);
+        assertThat(properties).contains(
+            entry(COURT_CODE_KEY, COURT_CODE),
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/service/TelemetryServiceTest.java:119]: Ah! Didn't realise you could match multiple entries with one assert! 👍

File: src/test/java/uk/gov/justice/probation/courtcasematcher/service/TelemetryServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+    @DisplayName("Simple record of event with no properties")
+    @Test
+    void whenMessageReceived_thenRecord() {
+        telemetryService.trackEvent(TelemetryEventType.COURT_LIST_RECEIVED);
+
+        verify(telemetryClient).trackEvent("PiCCourtListReceived");
+    }
+
+    @DisplayName("Record the event when an exact match happens")
```
Comment (jhevans) [inline src/test/java/uk/gov/justice/probation/courtcasematcher/service/TelemetryServiceTest.java:102]: This is more an observation than anything but the approach I took with the court-case-service TelemetryService was to test it in a way that's agnostic to the actual event being tracked. So as a service it's expected to be able to take an event type and an object and produce telemetry based on the contents of that object. The advantage being that you only really need to test two cases per method - the one where everything is present, and where the minimal subset is present. It's then up to the calling service to ensure that the right data is being sent, giving a clean separation of responsibilities.

**Review by jhevans — CHANGES_REQUESTED** [review]
There's one bean I think we can get rid of and I think there was one very minor suggestion but otherwise looks good

File: src/main/java/uk/gov/justice/probation/courtcasematcher/application/TelemetryConfig.java (hunk 1 above)
Comment (chris-faulkner-mojdt) [reply src/main/java/uk/gov/justice/probation/courtcasematcher/application/TelemetryConfig.java:40]: Done

File: src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/MessageProcessorTest.java (hunk 2 above)
Comment (chris-faulkner-mojdt) [reply src/test/java/uk/gov/justice/probation/courtcasematcher/messaging/MessageProcessorTest.java:124]: done

File: src/test/java/uk/gov/justice/probation/courtcasematcher/service/TelemetryServiceTest.java (hunk 4 above)
Comment (chris-faulkner-mojdt) [reply src/test/java/uk/gov/justice/probation/courtcasematcher/service/TelemetryServiceTest.java:102]: I see what you mean. I think that you have taken the approach of passing the type of event to the service, presumably making the decision about which type is required  outside the method.Presumably that has meant more complexity in the calling services ?

Comment (jhevans) [reply src/test/java/uk/gov/justice/probation/courtcasematcher/service/TelemetryServiceTest.java:102]: Yes marginally, though it's just one additional argument. 

What I hadn't clocked about this use case is that the objects being acted upon are different in each case - whereas in the court-case-service we have multiple events with the same object being logged so you gain a benefit from being able to reuse the methods. Given none of the methods here are reusable anyway it doesn't make much sense to do it the way I did it in court-case-service. 

I commented more out of curiosity than anything as I suspected there would be a good reason why we did it differently!

**Review by jhevans — APPROVED** [review]

**Pull request merged**